src/stavenote.ts

```diff
@@ -197,9 +197,85 @@ export class StaveNote extends StemmableNote {
           xShift = voiceXShift + 2;
           noteU.note.setXShift(xShift);
         } else {
-          // shift lower voice right
-          xShift = voiceXShift + 2;
-          noteL.note.setXShift(xShift);
+          //Vexflowpatch: Instead of shifting notes, remove the appropriate flag
+          //If we are sharing a line, switch one notes stem direction.
+          //If we are sharing a line and in the same voice, only then offset one note
+          const lineDiff = Math.abs(noteU.line - noteL.line);
+          if (noteU.note.hasStem() && noteL.note.hasStem()) {
+            //If we have different dot values, must offset
+            //Or If we have a non-filled in mixed with a filled in notehead, must offset
+            if (
+              noteU.note.duration !== noteL.note.duration ||
+              noteU.note.getModifiersByType('Dot').length !== noteL.note.getModifiersByType('Dot').length
+            ) {
+              xShift = voiceXShift + 2;
+              if (noteU.stemDirection === noteL.stemDirection) {
+                // upper voice is middle voice, so shift it right
+                noteU.note.setXShift(xShift);
+              } else {
+                // shift lower voice right
+                noteL.note.setXShift(xShift);
+              }
+              if (noteU.note.getModifiersByType('Dot').length > 0) {
```

Comment:
I think we can wrap this logic into the `Dot` class, so it continues to know where it is. I'd rather call `dot.shiftY()` maybe called by `note.shiftDotsUp(...)` or something.

Reply:
I have simplified the code a lot. I have also resolved a bug in `note`. The logic cannot be done within `Dot` because the lines of noteU and noteL are required.